KEYCLOAK-12001: Audience support for SAML clients

rmartinc opened the pull request:
Adding audience mappers for SAML. I have replicated the same behavior that OIDC has (even the labels). Comments:

* I used the (not used yet in any mapper) SAMLLoginResponseMapper but I modified the interface to pass the ClientSessionContext instead of the AuthenticatedClientSessionModel. As it was not used by any mapper and it's not documented I decided to change the interface.
* I reused all the labels in the mappers from the OIDC (only needed  in one of them) and replicated the same behavior.
* I created an audience resolve mapper although I don't think it was totally needed, but I preferred to implement the same mapper that exist in OIDC.

@hmlnarik As always if you want some change or you see any mistake or improvement just let me now. As in the other SAML PR there is no hurry at all, I'm just completing this in order to give a better SAML assertion replay experience.

rmartinc committed:
KEYCLOAK-12001: Audience support for SAML clients

hmlnarik commented on services/src/main/java/org/keycloak/protocol/saml/mappers/SAMLLoginResponseMapper.java:
```diff
@@ -29,6 +29,7 @@
+
     ResponseType transformLoginResponse(ResponseType response, ProtocolMapperModel mappingModel, KeycloakSession session,
-                                        UserSessionModel userSession, AuthenticatedClientSessionModel clientSession);
+                                        UserSessionModel userSession, ClientSessionContext clientSessionCtx);
```
Why is this change here?

rmartinc replied:
This change is because the audience resolve mapper needs to compute the groups of the user in each client (the resolve mapper duplicates the OIDC behavior, and the clientId is added in the audience if the user has at least one group in that client). So for that I re-used the [clientSessionCtx.getRoles()](https://github.com/keycloak/keycloak/pull/6520/files#diff-ec80ecf52bea4386f9a2e6e04c51d6b1R85) method. This class is also passed to the [SAMLRoleListMapper](https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/protocol/saml/mappers/SAMLRoleListMapper.java#L32). If I hadn't passed it I would have needed to calculate groups again (repeating code and calculation), as I saw no mapper used this interface I decided to modify it (the previous AuthenticatedClientSessionModel can also be accessed through the ClientSessionContext).

hmlnarik replied:
Thanks for explanation, it really makes sense to keep this on par with `OIDCAccessTokenMapper`.

hmlnarik reviewed (changes requested):
Thanks @rmartinc ! I realized I missed pushing "Submit review", sorry about that

rmartinc commented on services/src/main/java/org/keycloak/protocol/saml/mappers/SAMLAudienceResolveProtocolMapper.java:
```diff
@@ -0,0 +1,107 @@
+            logger.debugf("Calculated audiences to add: %s", audiences);
+            // add the audiences
+            for (String audience : audiences) {
+                aud.addAudience(URI.create(audience));
```
I have added a try block in both mappers.

rmartinc commented on services/src/main/java/org/keycloak/protocol/saml/mappers/SAMLAudienceProtocolMapper.java:
```diff
@@ -0,0 +1,136 @@
+        try {
+            return response.getAssertions().get(0).getAssertion().getConditions().getConditions()
+                    .stream()
+                    .filter(c -> c instanceof AudienceRestrictionType)
```
Done!

rmartinc commented:
Happy new year @hmlnarik! I added the first two modifications to the PR and explained why I changed the interface. Let me know if you prefer something different for the third point.

iankko commented:
@rmartinc This one seems to have broken current _master_, when built with the distribution (_-Pdistribution_) profile. Filed [KEYCLOAK-12865](https://issues.redhat.com/browse/KEYCLOAK-12865) to track this. Will you have a chance to look at this yet && propose a fix, or should we fix it?

Thank you!, Jan

iankko commented:
@rmartinc 

Created also [a PR for it](https://github.com/keycloak/keycloak/pull/6712) (seemed to be pretty straightforward). Could you review?

Thanks! Jan

> @rmartinc This one seems to have broken current _master_, when built with the distribution (_-Pdistribution_) profile. Filed [KEYCLOAK-12865](https://issues.redhat.com/browse/KEYCLOAK-12865) to track this. Will you have a chance to look at this yet && propose a fix, or should we fix it?
> 
> Thank you!, Jan

rmartinc commented:
> @rmartinc
> 
> Created also [a PR for it](https://github.com/keycloak/keycloak/pull/6712) (seemed to be pretty straightforward). Could you review?
> 
> Thanks! Jan
> 
> > @rmartinc This one seems to have broken current _master_, when built with the distribution (_-Pdistribution_) profile. Filed [KEYCLOAK-12865](https://issues.redhat.com/browse/KEYCLOAK-12865) to track this. Will you have a chance to look at this yet && propose a fix, or should we fix it?
> > Thank you!, Jan

The PR LGTM. I have even run that test in my env. Wait for the tests to pass. I have approved your PR but I don't know if I can do that. :-P

iankko commented:
@rmartinc 

> The PR LGTM. I have even run that test in my env. Wait for the tests to pass. I have approved your PR but I don't know if I can do that. :-P

Sure you can. Thanks for doing that & for re-running the necessary tests on your part! Just waiting for the Travis tests that one to finish OK too yet.
